File: haskell/004/Main.hs
{-
    ProjectEuler 004
    Haskell / Main.hs
    copyright (c) 2015 Susisu

    GHC 7.8.3
    `ghc Main.hs -o bin/main`
-}

module Main where

main :: IO ()
main = print $ calc 100 100 0

calc :: Int -> Int -> Int -> Int
calc x y p
    | x >= 1000 = p
    | y >= 1000 = calc (x + 1) 100 p
    | otherwise = let q = x * y in
        if isPalindrome q then
            calc x (y + 1) (max p q)
        else
            calc x (y + 1) p

isPalindrome :: (Integral a, Show a) => a -> Bool
isPalindrome n = let s = show n in s == reverse s

File: perl/004/main.pl
# ProjectEuler 004
# Perl / main.pl
# copyright (c) 2015 Susisu
#
# Perl 5.16.2
# `perl main.pl`

sub isPalindrome {
    my $s = $_[0] . "";
    $s == reverse $s;
}

sub main {
    my $max = 0;
    for (my $i = 100; $i < 1000; $i++) {
        for (my $j = 100; $j < 1000; $j++) {
            my $p = $i * $j;
            if (isPalindrome($p) && $p > $max) {
                $max = $p;
            }
        }
    }
    print $max . "\n";
}

&main;
